### ISSUE TITLE
Libbpf packaging of v0.8.x and v1.x series in Linux distros

### ISSUE THREAD

**Issue by anakryiko** [libbpf/libbpf#590]
I've created this issue as a form of communication with Linux distro packagers of libbpf. Hopefully that works OK as a medium.

As most people using libbpf probably know by now, we've been working on backwards-compatibility breaking major v1 release for a while (more than 18 months) now. As we went along with this, all the BPF and libbpf users were encouraged to adapt their code to use non-deprecated libbpf APIs and rely on libbpf 1.0 behaviors (https://github.com/libbpf/libbpf/wiki/Libbpf-1.0-migration-guide for an overview). To make it easier on users, we've been gradually adding behavior knobs (`libbpf_set_strict_mode()` API) and (whenever necessary and possible) replacement APIs in v0.x versions starting from v0.5. What this means in practice is that for majority of applications, somewhere starting from v0.6-v0.7 it was possible to have a code base completely compatible with both v0.x and v1.0 libbpf versions, both at the API level and in terms of strictness and supported set of features.

But this being a real life, it's not always possible for libbpf users to switch to libbpf v1.0 overnight. One such example is systemd which has a pretty long-tail backwards compatibility with older distros that don't provide libbpf 1.0. For such cases, the thinking was that distros would provide v0.x (essentially v0.8.1 and whatever bugfix patch versions we'll have afterwards, if any) for applications that can't yet switch to libbpf v1.0. And then for applications that don't have such a long-term support restrictions default "series" would follow v1.x releases.

This issue is to bring together libbpf package maintainers in various Linux distros to discuss this and agree on more-or-less consistent approach. Basically, as libbpf maintainer my request is to provide v1.x releases going forward, but also keep supplying v0.8.x versions for widest possible coverage for projects like systemd. I don't think such setup is unprecedented (I was pointed to libusb as one such example), even if maintaining two packages is more work than maintaining. But given there is probably not going much activity in v0.8.x series, I hope this will be mostly one-time setup for the same of allowing users more time to switch to v1.x.

Please let me know if this makes sense and if there is anything I as libbpf maintainer can help with this. Also please feel free to ping relevant people that I most certainly missed. Thank you!

cc @acj, @vt-alt, @sudipm-mukherjee, @olsajiri, @chutz, @jsmolic, @martinetd, @r-ryantm, @shunghsiyu

**Comment by martinetd** [comment]
Thanks for making this issue!

In my opinion the difficulty was that some of the replacement API were introduced progressively, so while users have been encouraged to stop using the older APIs e.g. systemd 1.0-compatible version depends on 0.7 which was released just half a year ago; that's a bit short for many distros, and applications can only get started after their maintainer's distro upgraded so even shorter there.

Anyway, for my hats:
 - alpine already released 1.0 and will just roll forward, we didn't have as many dependencies.
 - for nixos I'll add 1.0 as a separate package for now and switch users as we can as suggested; there is more than just systemd that didn't have support for 1.0; keeping both in parallel as suggested is probably the best way forward and we were already considering it.

**Comment by anakryiko** [comment]
> Thanks for making this issue!
> 
> In my opinion the difficulty was that some of the replacement API were introduced progressively, so while users have been encouraged to stop using the older APIs e.g. systemd 1.0-compatible version depends on 0.7 which was released just half a year ago; that's a bit short for many distros, and applications can only get started after their maintainer's distro upgraded so even shorter there.

I never tracked very precisely, but looking at https://repology.org/project/libbpf/versions periodically I got impression that major distros update libbpf versions in their distros pretty frequently and track latest releases quite closely overall. They don't always bother packaging bug-fix patch releases (e.g., some distros never got v0.8.1), which is sad, but I don't feel like I have much control here. BTW, I just released v1.0.1, I hope people will package that one over v1.0, because it fixes few annoyances reported after v1.0 release.

One big exception seems to be Ubuntu, which for some reason stuck to v0.5 while v0.7 was 2.5 months old by the time Ubuntu 22.04 was released.

As for gradual addition of replacement APIs, most replacements happened in v0.5 and v0.6, v0.7 cleaned up few feature-probing APIs (and systemd pretty much was the only reason we even reworked them instead of just dropping them, tbh; systemd is the only real user of them as far as I'm aware) and reworked XDP-related APIs (which I don't think systemd cares about), so absolute majority of applications would be 1.0-compatible as soon as v0.6. As much as I'd like for all this clean up to happen within single release super early on, reality is not as simple, unfortunately.

> 
> Anyway, for my hats:
> 
>     * alpine already released 1.0 and will just roll forward, we didn't have as many dependencies.
> 
>     * for nixos I'll add 1.0 as a separate package for now and switch users as we can as suggested; there is more than just systemd that didn't have support for 1.0; keeping both in parallel as suggested is probably the best way forward and we were already considering it.

Sounds good, thank you!

**Comment by anakryiko** [comment]
@acj, @vt-alt, @sudipm-mukherjee, @olsajiri, @chutz, @jsmolic, @r-ryantm, @shunghsiyu ping, any thoughts or objections?

**Comment by shunghsiyu** [comment]
Thanks for opening up this discussion! FWIW it may worth discuss different cases separately:

a. rolling-release distros (e.g. Arch)
b. snapshot-release distros that is still work-in-progress (e.g. Ubuntu 23.04)
c. snapshot-release distros that's out there already (e.g. Ubuntu 22.04)

But before that, I want to make sure I'm not misunderstanding anything, is the following points correct?

1. As of today systemd (v251) won't build with libbpf v1.0
2. v0.8.x series will continue to receive bug fix releases for some indefinite time
3. The suggestion is keep both v0.8.x and v1.0+ coexisting until the v0.8.x series is retired

---

Being quite inexperienced with packaging myself, I tried to looks for other examples of libraries with multiple version coexisting. The libusb case is actually quite interesting, it seems that libusb0 is not packaged anymore and is provided by [libusb-compat (on openSUSE Tumbleweed at least)](https://build.opensuse.org/package/show/openSUSE:Factory/libusb-compat), which itself uses libusb1 underneath.

From a distro perspective this seems easier to justify than having multiple version coexisting, but admittedly will require much more work. Any other pros/cons/thoughts?

**Comment by martinetd** [comment]
> As of today systemd (v251) won't build with libbpf v1.0

it won't have libbpf features, yes. systemd itself will work just fine, just a subset of optional features that will be disabled.
Discussions are still in progress but it'll -at least- work with libbpf 0.7+ or 1.0 (hopefully from next release), or possibly from 0.3+ depending on what maintainers decide for it. (The code is ready for both version, just need to come to an agreement and eventually fine-tune a bit but it should work)

> libusb-compat provided by libusb1

That definitely sounds possible, but as you pointed out someone needs to do the work. From a look at what breaks with libbpf 1.0 in the systemd ecosystem I'm expecting dependencies to be fixed quite quickly (in the order of months at most), so I don't think it's really worth the trouble if libbpf 0.8 keeps being maintained for long enough for distros to make the transition.
Taking your previous segmentation, I'd say:
 - rolling releases can just upgrade whenever all their dependencies are ready (e.g. arch already did afaik, systemd might not be compiled with bpf framework enabled?)
 -  snapshot releases in progress could package both if they want, but it's also probably ok to stay with 0.8 unless/until all their libbpf reverse dependencies work with libbpf 1. I assume the upgrade would be possible before the snapshot freeze for e.g. ubuntu 23.04 but that depends on each distro
 - snapshot releases already out should just keep their current version, or if they need new APIs upgrade to at most 0.8 for maximum compatibliity (this might depend on e.g. which path systemd decides on _IF_ they require systemd bpf features, from a quick check in https://github.com/systemd/systemd/pull/24507 it looks like most stable distros did not enable bpf in systemd so they just wouldn't care, but it looks like the data might not be accurate... Either way such distros would likely not be upgrading to the latest systemd either so my personal opinion on this is that it's overthinking, but I'm somewhat easygoing :P

**Comment by anakryiko** [comment]
> Thanks for opening up this discussion! FWIW it may worth discuss different cases separately:
> 
> a. rolling-release distros (e.g. Arch) b. snapshot-release distros that's still work-in-progress (e.g. Ubuntu 23.04) c. snapshot-release distros that's out there already (e.g. Ubuntu 22.04)
> 
> But before that, I want to make sure I'm not misunderstanding anything, is the following points correct?
> 
>     1. As of today systemd (v251) won't build with libbpf v1.0

as @martinetd mentioned, it will soon(-ish), systemd folks just need to decide on minimal 0.x version they are willing to dictate.
> 
>     2. v0.8.x series will continue to receive bug fix releases for some indefinite time

Yes, but so far in practice we did few bugfix releases only, and usually pretty soon after minor version. So in all likelihood there won't be any further updates to v0.8.1. But if something comes up requiring new release, then yes, there will be v0.8.2 and so on.

> 
>     3. The suggestion is keep both v0.8.x and v1.0+ coexisting until the v0.8.x series is retired

Yes, as long as applications like systemd and iproute2 (not sure what else would like/want to stick to v0.x series).

> From a distro perspective this seems easier to justify than having multiple version coexisting, but admittedly will require much more work. Any other pros/cons/thoughts?

I think the cons is for packagers to have to set up two packages, but for v0.8.x it should be one-time cost.

BTW, you mentioned Ubuntu 22.04 above. Do you know/think it would be possible to bump supported libbpf version in that LTS release to at least v0.7 (preferably all the way to v0.8.1)? I just don't know how this whole packaging and versioning works for distros and if already released Ubuntu LTSs can still upgrade their package versions.

Thanks!

**Comment by shunghsiyu** [comment]
> > 3. The suggestion is keep both v0.8.x and v1.0+ coexisting until the v0.8.x series is retired
> 
> Yes, as long as applications like systemd and iproute2 (not sure what else would like/want to stick to v0.x series).

Thanks for the clarifications. That would be really helpful.

> BTW, you mentioned Ubuntu 22.04 above. Do you know/think it would be possible to bump supported libbpf version in that LTS release to at least v0.7 (preferably all the way to v0.8.1)? I just don't know how this whole packaging and versioning works for distros and if already released Ubuntu LTSs can still upgrade their package versions.  

I can only speak for SLES and openSUSE (though RHEL/CentOS/Ubuntu/etc should have something along the same line), say if a package in openSUSE Leap 15.4 were to be upgraded (it's a released snapshot distro), it must not break any existing user applications, which means:  

  - API-compatibility  
  - ABI-compatibility, so no changes in
  	- function call signature  
  	- user-visible structs (either directly or indirectly)  
  - and other bits and pieces  
  	- e.g. if build pipeline break due to deprecation warning then that may be considered as a breakage  

The above points are hard to guarantee during package upgrade, so in practice we snapshot distro packager tends to just stick to the version use during release (or whatever bug-fix release of it) and backport/cherry-pick fixes as need until EOL.

In short, I don't think libbpf in released snapshot distro cannot be easily upgraded. Maintaining libbpf v1.0 alongside (which you've proposed) is far easier than an upgrade, but for snapshot distro there isn't much incentive to do that.

A more approachable way of bringing libbpf v1.0 to Ubuntu LTS and the likes is to have it hosted in an independent package repository. The Open Build Service may help if that route is taken (e.g [OpenSCAD uses openSUSE's Open Build Service](https://build.opensuse.org/repositories/home:t-paul/OpenSCAD) to [build its package for Debian, Fedora, Ubuntu, etc. besides openSUSE](https://openscad.org/downloads.html)).

**Comment by shunghsiyu** [comment]
> * rolling releases can just upgrade whenever all their dependencies are ready (e.g. arch already did afaik, systemd might not be compiled with bpf framework enabled?)
> * snapshot releases in progress could package both if they want, but it's also probably ok to stay with 0.8 unless/until all their libbpf reverse dependencies work with libbpf 1 ...

Agree with the above. (and also wondering whether Arch users ran into problem with systemd)

> * snapshot releases already out should just keep their current version, or if they need new APIs upgrade to at most 0.8 for maximum compatibliity...

I'd say it quite unlikely to see snapshot release that already out to upgrade libbpf [for reasons mentioned above](https://github.com/libbpf/libbpf/issues/590#issuecomment-1268448947). Most if not all would probably just keep their current version. (At least that's my plan for openSUSE Leap)

**Comment by anakryiko** [comment]
> > BTW, you mentioned Ubuntu 22.04 above. Do you know/think it would be possible to bump supported libbpf version in that LTS release to at least v0.7 (preferably all the way to v0.8.1)? I just don't know how this whole packaging and versioning works for distros and if already released Ubuntu LTSs can still upgrade their package versions.
> 
> I can only speak for SLES and openSUSE (though RHEL/CentOS/Ubuntu/etc should have something along the same line), say if a package in openSUSE Leap 15.4 were to be upgraded (it's a snapshot releases that's out), it must not break any existing user applications, which means:
> 
>     * API-compatibility
> 
>     * ABI-compatibility, so no changes in
>       
>       * function call signature
>       * user-visible structs (either directly or indirectly)
> 
>     * and other bits and pieces
>       
>       * e.g. if build pipeline break due to deprecation warning then that may be considered as a breakage
> 
> 
> The above points are hard to guarantee during package upgrade, so in practice we packager working with snapshot releases tends to just stick to the version used during release (or whatever bug-fix release of it) and backport/cherry-pick fixes as need until EOL.
> 

So this was a very explicit and conscious corner stone of minor releases of libbpf: to never break both API and ABI compatibility. Barring any uncaught bugs, there should be no ABI breakage between any v0.x releases. So I think it's pretty safe to bump minor version.

> In short, I don't think libbpf in snapshot released cannot be easily upgraded. Maintaining libbpf v1.0 alongside (which you've proposed) is far easier than an upgrade, but for snapshot releases there isn't much incentive to do that.
> 

See above. API/ABI backwards and forward compat was always on libbpf's mind. So if that's the only reason holding anyone from upgrading libbpf 0.x release, I'd advocate to not hesitate with upgrade.

**Comment by shunghsiyu** [comment]
> So this was a very explicit and conscious corner stone of minor releases of libbpf: to never break both API and ABI compatibility. Barring any uncaught bugs, there should be no ABI breakage between any v0.x releases. So I think it's pretty safe to bump minor version.

Kudos for the API and ABI compatibility across v0.x (I had thought that libbpf follows the usual semantic versioning where v0.x is not considered stable API). Given the compatibility I agree it should be safe to bring v0.x up to v0.8.1, let me try to initiate such process for SLES and openSUSE Leap then.

**Comment by jameshilliard** [comment]
I maintain libbpf/bpftool in [buildroot](https://buildroot.org/) and generally I've been posting patches to master shortly after each release(and we have fairly short release cycles, however we don't have many packages consuming libbpf yet so we haven't had much in the way of versioning issue).

I suspect in the future the main concern for us will be more toolchain/kernel version compatibility related than anything as we tend to support building with a few older toolchain release versions for GCC at least(for example we currently support 10.x 11.x and 12.x series) as well as many kernel versions(in addition to allowing users to build with external kernel trees).

Our llvm toolchain support is much more limited(largely due to supporting far fewer architectures than llvm) and we generally only support one version there(and only support it for a few things where GCC builds aren't feasible).

We probably won't be able to build bpf programs with GCC however for most applications like systemd until GCC 13.x is released but once we have that working as long as we don't hit toolchain related regressions in new libbpf releases(which I haven't seen so far) then the main thing is just not breaking API compatibility with applications using libbpf in our tree.

Since we tend to build everything from source API compatibility tends to be more important for us than ABI compatibility as we effectively rebuild applications when their dependencies change most of the time so we don't typically even notice ABI compatibility breaks.
